Build failure: python313Packages.pybind11 (`armv7l` cross)

### Nixpkgs version

- Stable (25.11)

### Steps to reproduce

Run the following from the local nixpkgs repo (branch: nixos-25.11):
```sh
nix build .#pkgsCross.armv7l-hf-multiplatform.python313Packages.pybind11 -L
```

### Can Hydra reproduce this build failure?

Please select the Hydra Status.

### Link to Hydra build job

_No response_

### Relevant log output

```console
python3.13-pybind11> ERROR Backend subprocess exited when trying to invoke build_wheel
error: builder for '/nix/store/aaz9cpwsb5mjm6qhsbdnlw8wbwm2ma7h-python3.13-pybind11-3.0.1-armv7l-unknown-linux-gnueabihf.drv' failed with exit code 1;
       last 25 log lines:
       >                  from /build/source/include/pybind11/eval.h:14,
       >                  from /build/source/tests/pybind11_tests.h:3,
       >                  from /build/source/tests/test_class_sh_basic.cpp:1:
       > /nix/store/3lll9y925zz9393sa59h653xik66srjb-python3-3.13.9/include/python3.13/pyport.h:399:2: error: #error "LONG_BIT definition appears wrong for platform (bad gcc/glibc config?)."
       >   399 | #error "LONG_BIT definition appears wrong for platform (bad gcc/glibc config?)."
       >       |  ^~~~~
       > [36/108] Building CXX object tests/CMakeFiles/pybind11_tests.dir/test_class.cpp.o
       > FAILED: [code=1] tests/CMakeFiles/pybind11_tests.dir/test_class.cpp.o
       > /nix/store/br8a01761nsa7kpr42i6dd9sdmk78xi1-armv7l-unknown-linux-gnueabihf-gcc-wrapper-14.3.0/bin/armv7l-unknown-linux-gnueabihf-g++ -DPYBIND11_TEST_BOOST -DPYBIND11_TEST_EIGEN -Dpybind11_tests_EXPORTS -I/build/source/include -isystem /nix/store/3lll9y925zz9393sa59h653xik66srjb-python3-3.13.9/include/python3.13 -isystem /nix/store/m20sxgg1xyxjwy5df9b1gajwspzhjpzz-eigen-armv7l-unknown-linux-gnueabihf-3.4.1/include/eigen3 -O3 -DNDEBUG -fPIC -fvisibility=hidden -Wall -Wextra -Wconversion -Wcast-qual -Wdeprecated -Wundef -Wnon-virtual-dtor -flto=auto -fno-fat-lto-objects -MD -MT tests/CMakeFiles/pybind11_tests.dir/test_class.cpp.o -MF tests/CMakeFiles/pybind11_tests.dir/test_class.cpp.o.d -o tests/CMakeFiles/pybind11_tests.dir/test_class.cpp.o -c /build/source/tests/test_class.cpp
       > In file included from /nix/store/3lll9y925zz9393sa59h653xik66srjb-python3-3.13.9/include/python3.13/Python.h:64,
       >                  from /build/source/include/pybind11/conduit/wrap_include_python_h.h:44,
       >                  from /build/source/include/pybind11/detail/common.h:12,
       >                  from /build/source/include/pybind11/attr.h:13,
       >                  from /build/source/include/pybind11/detail/class.h:12,
       >                  from /build/source/include/pybind11/pybind11.h:12,
       >                  from /build/source/include/pybind11/stl.h:12,
       >                  from /build/source/tests/test_class.cpp:17:
       > /nix/store/3lll9y925zz9393sa59h653xik66srjb-python3-3.13.9/include/python3.13/pyport.h:399:2: error: #error "LONG_BIT definition appears wrong for platform (bad gcc/glibc config?)."
       >   399 | #error "LONG_BIT definition appears wrong for platform (bad gcc/glibc config?)."
       >       |  ^~~~~
       > ninja: build stopped: subcommand failed.
       > 
       > *** CMake build failed
       >
       > ERROR Backend subprocess exited when trying to invoke build_wheel
       For full logs, run:
         nix log /nix/store/aaz9cpwsb5mjm6qhsbdnlw8wbwm2ma7h-python3.13-pybind11-3.0.1-armv7l-unknown-linux-gnueabihf.drv
```

### Additional context

_No response_

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.18.0, NixOS, 25.11 (Xantusia), 25.11.20251206.d9bc5c7`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.31.2`
 - nixpkgs: `/nix/store/mmbrs76ksccdn0bvvqs4plivwh6xlcs5-source`

### Notify maintainers

@yuriaisaka @dotlambda

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/python-modules/pybind11/default.nix
+++ b/pkgs/development/python-modules/pybind11/default.nix
@@ -14,6 +14,9 @@
   numpy,
   pytest,
   makeSetupHook,
+  # Build tests to verify cross-compilation works, but only when CPU bit
+  # depth matches (otherwise Python headers cause LONG_BIT mismatch errors)
+  buildTests ? stdenv.hostPlatform.parsed.cpu.bits == stdenv.buildPlatform.parsed.cpu.bits,
 }:
 let
   setupHook = makeSetupHook {
@@ -44,9 +47,7 @@ buildPythonPackage rec {
     pybind11.passthru.scikit-build-core-no-tests
   ];
 
-  buildInputs = [
-    # Used only for building tests - something we do even when cross
-    # compiling.
+  buildInputs = lib.optionals buildTests [
     catch2
     boost
     eigen
@@ -65,13 +66,11 @@ buildPythonPackage rec {
   ];
 
   cmakeFlags = [
-    # Always build tests, because even when cross compiling building the tests
-    # is another confirmation that everything is OK.
-    (lib.cmakeBool "BUILD_TESTING" true)
+    (lib.cmakeBool "BUILD_TESTING" buildTests)
 
     # Override the `PYBIND11_NOPYTHON = true` in `pyproject.toml`. This
     # is required to build the tests.
-    (lib.cmakeBool "PYBIND11_NOPYTHON" false)
+    (lib.cmakeBool "PYBIND11_NOPYTHON" (!buildTests))
   ];
 
   dontUseCmakeConfigure = true;

